src/batch/src/task/context.rs

```diff
@@ -67,6 +68,49 @@ pub trait BatchTaskContext: Clone + Send + Sync + 'static {
     fn mem_usage(&self) -> usize;
 
     fn create_executor_mem_context(&self, executor_id: &str) -> Option<MemoryContextRef>;
+    fn get_stop_flag(&self) -> Arc<StopFlag>;
+
+    fn get_stop_flag_ref(&self) -> &Arc<StopFlag>;
+}
+
+/// We already have `shut_down` sender to stop execution.
+/// `StopFlag` used in executor to let it stop execute more timely when it can't receive message
+/// timely.
+#[derive(Clone)]
+pub struct StopFlag {
+    stop_flag: Arc<AtomicBool>,
+    stop_reason: Arc<RwLock<Option<String>>>,
+}
+
+impl StopFlag {
+    pub fn new() -> Self {
+        Self {
+            stop_flag: Arc::new(AtomicBool::new(false)),
+            stop_reason: Arc::new(RwLock::new(None)),
+        }
+    }
+
+    pub fn set_stop(&self, reason: String) {
+        self.stop_reason.write().replace(reason);
+        // NOTE: Used `store(Ordering::Release),load(Ordering::Acquire)` to guarantee that the stop
+        // reason can be viewed if the stop flag is visible.
+        self.stop_flag.store(true, Ordering::Release);
+    }
+
+    pub fn check_stop(&self) -> Option<String> {
```

Comment:
Why this should return an `Option<String>`, I think a bool would be enough?

Reply:
Because we need the error reason when the executor stop with StopFlag.
```
fn do_execute() {
    if Some(error_reason) = self.stop_flag.check_stop() {
        return Err(error_reason.into());
    }
}
```

Reply:
Hmm, we don't need to do that, we can check this in `run` method:
https://github.com/risingwavelabs/risingwave/blob/337b7e3279bca9453b20e1938b13d0eca576def0/src/batch/src/task/task_execution.rs#L496